eventbus/kafka/sensor/kafka_sensor.go

```diff
@@ -0,0 +1,355 @@
+	return !s.connected || s.client.Closed()
+}
+
+func (s *KafkaSensor) Event(msg *sarama.ConsumerMessage) ([]*sarama.ProducerMessage, int64, func()) {
+	var event *cloudevents.Event
+	if err := json.Unmarshal(msg.Value, &event); err != nil {
+		s.Logger.Errorw("Failed to deserialize cloudevent, skipping", zap.Error(err))
+		return nil, msg.Offset + 1, nil
+	}
+
+	messages := []*sarama.ProducerMessage{}
+	for _, trigger := range s.triggers.List(event) {
+		event, err := trigger.Transform(trigger.depName, event)
+		if err != nil {
+			s.Logger.Errorw("Failed to transform cloudevent, skipping", zap.Error(err))
+			continue
+		}
+
+		if !trigger.Filter(trigger.depName, event) {
+			continue
+		}
+
+		// if the trigger only requires one message to be invoked we
+		// can skip ahead to the action topic, otherwise produce to
+		// the trigger topic
+
+		var data any
+		var topic string
+		if trigger.OneAndDone() {
+			data = []*cloudevents.Event{event}
+			topic = s.topics.action
+		} else {
+			data = event
+			topic = s.topics.trigger
+		}
+
+		value, err := json.Marshal(data)
+		if err != nil {
+			s.Logger.Errorw("Failed to serialize cloudevent, skipping", zap.Error(err))
+			continue
+		}
+
+		messages = append(messages, &sarama.ProducerMessage{
+			Topic: topic,
+			Key:   sarama.StringEncoder(trigger.Name()),
+			Value: sarama.ByteEncoder(value),
+		})
+	}
+
+	return messages, msg.Offset + 1, nil
+}
+
+func (s *KafkaSensor) Trigger(msg *sarama.ConsumerMessage) ([]*sarama.ProducerMessage, int64, func()) {
+	var event *cloudevents.Event
+	if err := json.Unmarshal(msg.Value, &event); err != nil {
+		s.Logger.Errorw("Failed to deserialize cloudevent, skipping", zap.Error(err))
+	}
+
+	messages := []*sarama.ProducerMessage{}
+	offset := msg.Offset + 1
+
+	// Update trigger with new event and add any resulting action to
+	// transaction messages
+	if trigger, ok := s.triggers[string(msg.Key)]; ok && event != nil {
+		func() {
+			events, err := trigger.Update(event, msg.Partition, msg.Offset, msg.Timestamp)
+			if err != nil {
+				s.Logger.Errorw("Failed to update trigger, skipping", zap.Error(err))
+				return
+			}
+
+			// no events, trigger not yet satisfied
+			if events == nil {
+				return
+			}
+
+			value, err := json.Marshal(events)
+			if err != nil {
+				s.Logger.Errorw("Failed to serialize cloudevent, skipping", zap.Error(err))
+				return
+			}
+
+			messages = append(messages, &sarama.ProducerMessage{
+				Topic: s.topics.action,
+				Key:   sarama.StringEncoder(trigger.Name()),
+				Value: sarama.ByteEncoder(value),
+			})
+		}()
+	}
+
+	// Need to determine smallest possible offset against all
+	// triggers as other triggers may have messages that land on the
+	// same partition
+	for _, trigger := range s.triggers {
+		offset = trigger.Offset(msg.Partition, offset)
+	}
+
+	return messages, offset, nil
+}
+
+func (s *KafkaSensor) Action(msg *sarama.ConsumerMessage) ([]*sarama.ProducerMessage, int64, func()) {
```

Comment:
Looks like for event/action/trigger the second return value is always msg.offset + 1. Maybe remove this as a return value if it's always the same

Reply:
It's not always the same, the trigger is the odd one out, it is set to the smallest required offset across all triggers. That value is set [here](https://github.com/intuit-data-os/argo-events/pull/10/files/690d42a83591d487dd4b2359f0f85b5444dd7d9f#diff-b778225ecfb035ee9a44077f3f987fd20bc3381f1786485f86639419b0c9196fR336)